### DISCUSSION TITLE
RAM usage is very high for a large dataset, is it normal ?

### DISCUSSION THREAD

**Discussion by RemyXia** [milvus-io/milvus#36345] [Q&A and General discussion]
Hello, I want to query over 5 collections of 20M vectors with 1024 dimensions. 
Right now I'm testing with 2M vectors and I see that the query speed is very good but my RAM usage is exceeding 10Go, therefore I tried offloading the indexes to the disk with mmap but the performance dropped significantly. 

I went from a few ms to more than a second for a single query. 
Is there a way to optimize RAM usage ? I'm new to this so I'm only familiar with non vector db and I know that indexes normally aren't that heavy but is it different with vectors ?

Thanks in advance for your help !

**Comment by yhmo** [comment]
You can estimate the original data size in this way: each dimension of vector is represented as a float32 value, which occupies 4 bytes. So, the original data size is: row_count * dimension * 4 bytes

For 1024 dim and 2M vectors, the original data size is 1024 * 4 * 2M = 8192MB = 8GB

Different index type has different in-memory sizes.
Typically, the index size compare to the original vectors size:
- IVF_FLAT/HNSW, a bit larger than the original vectors size
- IVF_SQ8/IVF_PQ, 25% ~ 30% of the original vectors size, the recall rate of IVF_PQ is not good.
- DISKANN, 1/4 ~ 1/6 of the original vectors size, performance is not good, requires high-performance hard disk.

So, if you choose IVF_FLAT, the in-memory size will be a bit larger than 8GB. There are extra memory cost for runtime, so 10GB is expected.
To save memory usage, you can use IVF_SQ8 or DISKANN.

Memory-mapped files enable efficient data handling by mapping file contents directly into memory. This is especially useful when memory is limited and loading all data is not possible. This technique can boost data capacity and maintain performance to a point. However, if the data greatly exceeds memory capacity, search and query speeds could significantly decrease. 

If you need high-speed search performance, it is recommended to use full in-memory index.
